src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs

```diff
@@ -215,17 +215,59 @@ private static unsafe nuint GetIndexOfFirstNonAsciiByte_Default(byte* pBuffer, n
             goto Finish;
         }
 
-        private static unsafe nuint GetIndexOfFirstNonAsciiByte_Sse2(byte* pBuffer, nuint bufferLength)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static ulong GetMaskOfMostSignificantBits(Vector128<byte> value)
```

Comment:
>GetMaskOfMostSignificantBits [](start = 30, length = 28)

Alright, so on careful inspection it looks like you can optimize this method for `byte` variant (all the places in `GetIndexOfFirstNonAsciiByte_Sse2OrArm64()` that calls this method). This is essentially 2nd use-case in my comment here - [https://github.com/dotnet/runtime/pull/38653#discussion_r453059892](https://github.com/dotnet/runtime/pull/38653#discussion_r453059892). For the `mask` that you get for `byte` version, you just do `TrailingZeroCount(mask)` on it. So you don't care about the exact mask but the first lane that has MSB set. So you can achieve the same thing as done in [https://github.com/dotnet/runtime/pull/38707](https://github.com/dotnet/runtime/pull/38707).

```csharp
[MethodImpl(MethodImplOptions.AggressiveInlining)]
public static int GetIndexOfFirstNonAsciiByte(Vector128<byte> value)
{
    if (!AdvSimd.Arm64.IsSupported || !BitConverter.IsLittleEndian)
    {
        throw new PlatformNotSupportedException();
    }

    Vector128<byte> mostSignificantBitIsSet = AdvSimd.ShiftRightArithmetic(value.AsSByte(), 7).AsByte();
    Vector128<byte> extractedBits = AdvSimd.And(mostSignificantBitIsSet, s_bitmask);

    // collapse mask to lower bits
    extractedBits = AdvSimd.Arm64.AddPairwise(extractedBits, extractedBits);
    ulong mask = extractedBits.AsUInt64().ToScalar();

    // calculate the index
    int index = BitOperations.TrailingZeroCount(mask) >> 2;
    Debug.Assert((mask != 0) ? index < 16 : index >= 16);
    return index;
}
```

 I would also rename the method used in `byte` version to `GetIndexOfFirstNonAsciiByte()`.

Reply:
You can optimize this method for `char` variant as well (all the places in `GetIndexOfFirstNonAsciiChar_Sse2OrArm64()` that calls this method). Even this falls in 2nd category where you just use the returned value and call `TrailingZeroCount()` on it. (Just double check my observation). So here is my thinking.

I am pasting @GrabYourPitchforks 's response just to have all this info in single place:

" `0xAAAA` is correct. The way this logic works is that it adds `0x7F80` to every 16-bit `ushort` in the input, saturating at `0xFFFF` (to avoid integer overflow back to a negative value).

This means that if the original value is in the range `0x0000 .. 0x007F` (ASCII), the add operation will result in `0x7F80 .. 0x7FFF`.
If the original value is in the range `0x0080 .. 0x807F` (non-ASCII), the add operation will result in `0x8000 .. 0xFFFF`.
And if the original value is in the range `0x8080 .. 0xFFFF` (non-ASCII), the add operation will result in `0xFFFF` due to saturation.

The non-ASCII case _always_ results in the sum having the `0x8000` bit set, but it doesn't always result in the sum having the `0x0080` bit set. Hence `0xAAAA` - not `0xFFFF` - is the correct comparison against the _pmovmskb_ result.

This is also called out in the comment a few lines above, which should hint that alternating bits of the _pmovmskb_ result should be ignored."

If you have non-Ascii value between `0x80 .. 0xFFFF`, the `AddSaturate()` operation will give you a value between `0x8000 ... 0xFFFF`. Then you can do similar trick as above, except do an arithmetic right shift by `15`. That way, you end up getting `0xFFFF` in all lanes that originally had values between `0x80...0xFFFF`. After that you do `And` with slightly modified `s_bitMask` to get the mask and from that get trailing zero count to get the lane. If all lanes contain ASCII, then the trailing count would 8 and thats the amount you need to advance the buffer pointer.

```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
public static int GetIndexOfFirstNonAsciiUShort(Vector128<ushort> value)
{
    if (!AdvSimd.Arm64.IsSupported || !BitConverter.IsLittleEndian)
    {
        throw new PlatformNotSupportedException();
    }

    // This can be readonly static
    Vector128<ushort> s_bitmask = Vector128.Create((uint)0x1000001).AsUInt16();
    
    Vector128<ushort> mostSignificantBitIsSet = AdvSimd.ShiftRightArithmetic(value.AsInt16(), 15).AsUInt16();
     Vector128<ushort> extractedBits = AdvSimd.And(mostSignificantBitIsSet, s_bitmask);
    // collapse mask to lower bits
     extractedBits = AdvSimd.Arm64.AddPairwise(extractedBits, extractedBits);
    ulong mask = extractedBits.AsUInt64().ToScalar();
    // calculate the index
    int index = BitOperations.TrailingZeroCount(mask) >> 3;
    Debug.Assert((mask != 0) ? index < 8 : index >= 8);
    return index;
}
```

With this suggestion, the callers of this method that pass the `value` as a result of `Saturate()` has to change:
1. It no longer need to case to `.AsByte()` but can simply do `GetIndexOfFirstNonAsciiUShort(AdvSimd.AddSaturate(firstVector, asciiMaskForAddSaturate))`.

2. You no longer have to do the check of `currentMask` returned from this method with `NonAsciiDataSeenMask`. So your AdvSimd won't have the check for `if ((currentMask & NonAsciiDataSeenMask) != 0)`. Instead it will be
```csharp
if (indexOfNonAsciiUShort < 8) { // renamed currentMask to indexOfNonAsciiUShort
   goto FoundNonAsciiDataInCurrentMask;
}
```

3. In the label `FoundNonAsciiDataInCurrentMask`, you no longer have to do `currentMask &= NonAsciiDataSeenMask;` and `BitOperations.TrailingZeroCount(currentMask) - 1`, but instead can directly do `pBuffer = (char*)((byte*)pBuffer + indexOfNonAsciiUShort);`

4. Update assert for this case to be `Debug.Assert(indexOfNonAsciiUShort  >= 8)`.

By doing this, you save 2 `AddPairwise` instructions in this operation. But please double check the functionality with various inputs to make sure the logic is correct. I might have missed something in the snippet I gave.

Reply:
Well, in point 3 above, inside `FoundNonAsciiDataInCurrentMask` , the buffer is byte* advanced with bytes factor, so you might want to do similar when you use `indexOfNonAsciiUShort` in the arithmetic.